### ISSUE TITLE
First build fails when used as a subdirectory in cmake build

### ISSUE THREAD

**Issue by JonnyPtn** [warmcat/libwebsockets#850]
Hi, I'm having a problem with including libwebsockets as part of my projects cmake build process.

I'm cloning the repository then adding it as a subdirectory, like so: 

`if(NOT LIBWEBSOCKETS_FOUND)
    execute_process(COMMAND git clone https://github.com/warmcat/libwebsockets ${CMAKE_SOURCE_DIR}/ThirdParty/websockets )
    add_subdirectory(ThirdParty/websockets ${CMAKE_SOURCE_DIR}/ThirdParty/websockets)
endif()`

But on the first build after building websockets_shared, I get the error `Error copying file "/home/jonny/workspace/project/lib/libwebsockets.h" to "/home/jonny/workspace/project/cmakebuild/include/libwebsockets.h".`

I checked the path and don't see the file in this location.Subsequent builds seem to work correctly, even though the file still doesn't seem to exist there.

I understand this probably only affects a small number of people (potentially just me?!), so thanks in advance for any help

Update: Anyone able to enlighten me as to the reason for these copy commands? Removing them hasn't broken anything in my preliminary tests

**Comment by JoakimSoderberg** [comment]
I replied in your PR that you closed: https://github.com/warmcat/libwebsockets/pull/849#issuecomment-289931254

But I realize this might be a better place to write. So copy paste:

Hello...

Now I have not tested this with the absolute latest version of libwebsockets, so it might be broken.

But 3 different ways to link with libwebsockets (with options).

1. Use CMake to download libwebsockets via git, build it and then link with the resulting binaries.

2. Add libwebsockets as a submodule to the project (or just dump the files if that floats your boat) into the project, and then point to that location for the CMake find_package command. Note this requires that you manually build the version first

3. Find the system version of libwebsockets and link with that.

(4. Another option would be to use the CMake add_subdirectory command, but not sure if the libwebsockets cmake project can handle that. Maybe this was what you wanted to do?)

```cmake
option(AUTO_LWS "Automatically download and build dependencies" OFF)
option(INTERNAL_LWS "Use the included version of Libwebsockets" ON)

# So in the example below you would build LWS using CMake and output the binaries to this path.
# (used for option 2, when INTERNAL_LWS is ON and AUTO_LWS is OFF)
set(LIBWEBSOCKETS_PATH "${PROJECT_SOURCE_DIR}/build_3rdparty/libwebsockets" CACHE PATH "Location of libwebsockets build tree")

# In my project I fetch the same commit as specified in the submodule for libwebsocket here
# when using option 2. But this can be hard coded to some commit or tag as well of course.
# (used for option 1, when AUTO_LWS is ON)
set(LIBWEBSOCKETS_TAG "2.2" CACHE STRING "Libwebsockets commit hash to build")

....

if (AUTO_LWS)
   # This method I use when cross compiling usually, since then all dependencies
   # can be built in one go using the exact same CMake toolchain file
   # that is passed here to the subproject...

   if (CMAKE_TOOLCHAIN_FILE)
    set(EXTERNAL_TOOLCHAIN "-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE}")
   endif()

    include(ExternalProject)

    # This makes CMake checkout the project using git, and then build it
    # using the same cmake generator as used to build the main project
    ExternalProject_Add(libwebsockets-git
        GIT_REPOSITORY "https://github.com/warmcat/libwebsockets.git"
        GIT_TAG "${LIBWEBSOCKETS_TAG}"
        CMAKE_GENERATOR "${CMAKE_GENERATOR}"
        CMAKE_ARGS ${EXTERNAL_TOOLCHAIN} ${EXTRA_EXTERNAL_ARGS} -DLWS_SSL_CLIENT_USE_OS_CA_CERTS=ON -DLWS_WITHOUT_TESTAPPS=ON -DCMAKE_BUILD_TYPE=MinSizeRel
        INSTALL_COMMAND ""
        )

    add_dependencies(theprogram libwebsockets-git)

    ExternalProject_Get_Property(libwebsockets-git SOURCE_DIR)
    ExternalProject_Get_Property(libwebsockets-git BINARY_DIR)

    set(LIBWEBSOCKETS_INCLUDE_DIRS
        "${SOURCE_DIR}/lib/"
        "${BINARY_DIR}")

    find_package(ZLIB REQUIRED)

    set(LIBWEBSOCKETS_STATIC_LIBRARIES "${BINARY_DIR}/lib/${CMAKE_CFG_INTDIR}/libwebsockets.a" ${ZLIB_LIBRARIES})

elseif (INTERNAL_LWS)
    # Find the version built in the project dir.
    find_package(Libwebsockets REQUIRED CONFIG NO_DEFAULT_PATH NO_CMAKE_PACKAGE_REGISTRY
                PATHS ${LIBWEBSOCKETS_PATH})

    # TODO: Move these to Libwebsockets itself.
    set(LIBWEBSOCKETS_STATIC_LIBRARIES websockets)
    set(LIBWEBSOCKETS_DYNAMIC_LIBRARIES websockets_shared)
else()
    # Find system version
    find_package(Libwebsockets REQUIRED)
endif()

# Finally we can use it.
include_directories(${LIBWEBSOCKETS_INCLUDE_DIRS})
target_link_libraries(theprogram ${LIBWEBSOCKETS_STATIC_LIBRARIES})
```

**Comment by lws-team** [comment]
First, generally, I don't think what you're trying to do there by cloning lws from CMake is a Good Idea.

This should be handled by git submodules... that is not a one-off solution like your clone-only plan above, but allows you to pin and track the remote repo in an ongoing way, managed from your meta-project git but without storing the submodule there.

https://git-scm.com/book/en/v2/Git-Tools-Submodules

The reason for the copies is when lws is integrated in a larger project, sibling projects may need to consume apis from lws without any formal system install, eg, the larger project may statically link, and / or not be targeting the system OS or CPU, and they may want to do so before lws gets built.  `./lib/libwebsockets.h` holds the public apis for lws, but it also requires the cmake `lws_config.h` to build... the first is part of the source bundle but the second is generated at CMake-time according to your options.

To give the other code in your meta-project somewhere to -I cleanly, lws CMake copies both libwebsockets.h and the generated lws_config.h in ./include in the lws build dir.  This gives the other projects the fixed dir to add with -I on their CFLAGS to access lws exports cleanly even before lws itself may have been built.

If there's something about that which makes trouble in your case, I'd like to fix it, however it works find here in other cases where lws is included in a larger project.

**Comment by JonnyPtn** [comment]
One more quickie - Is there a reason find_package_handle_standard_args isn't used in Findlibwebsockets.cmake? It effectively does the same as the block you have on lines 22-27 currently, but will actually output whether it has been found or not (currently there is no cmake output when finding lws), e.g.:

`-- Could NOT find LibWebSockets (missing:  LIBWEBSOCKETS_INCLUDE_DIR LIBWEBSOCKETS_LIBRARIES) `

I've made the change here, so if you are interested I can put together a PR straight away

**Comment by lws-team** [comment]
Hm... thumbs down for answering your questions... well, as I said cloning a subproject in the cmake seems like a shitty way to do things.

If you chose to use svn I'll leave it up to you how to work around not being able to use git submodules.

The copies are there because other use-cases need them as explained...

**Comment by JonnyPtn** [comment]
@lws-team And is that your official response? I'd rather not have to come down to that level

**Comment by lws-team** [comment]
Good luck bro!

**Comment by JonnyPtn** [comment]
@lws-team thanks for your mature, helpful approach to my problem. You are the absolute embodiment of everything that's good about open-source communities

**Comment by lws-team** [comment]
Yeah... I think so, thanks.

Not only do you get to use the result of years of my (+/- 130 or so contributors) work for free, I got up thismorning and answered your issue carefully and at length, despite nobody is paying me to do it.  I think I am living the FOSS life pretty nicely.

The key words there are "your problem".  If you expect me to throw more of my time after "your problem", it's advisable not to respond to help as you have done on this issue.  Because from a starting point of neutral willingness to assist you, you have inculcated a feeling best described as, "good luck, bro!"

**Comment by JonnyPtn** [comment]
I don't expect you to solve my problem, I don't expect you to modify anything you use to accomodate my methods, I don't expect you to do anything for me that you don't want to. I came here asking for help with a problem i have. I got one excellent, wonderful response which I couldn't have asked for more from (and in fact your code snippet helped me to find an alternative way around this @JoakimSoderberg, so thanks once more). @lws-team your response mostly told me that the way I was doing it was stupid, the way you're doing it works, and the problem exists to accomodate the way you and some others use the software. 

You can get righteous all you want, but you haven't been the slightest bit helpful. You can moan about my attitude all you want, but all I did was "thumbs down" an answer which wasn't helpful to me, which I always thought was the intended use of the responses.

Bottom line - I got two responses to my problem, one was fantastic, one was awful. I politely, neutrally, and calmly responded to both answers. If you're this offended by an emoji, best to stay off the internet.
